contracts/ERC20Safe.sol

```diff
@@ -165,26 +225,42 @@ contract ERC20Safe is BridgeRole, Pausable {
         batch.depositsCount++;
         depositsCount++;
 
-        tokenBalances[tokenAddress] += amount;
-
-        IERC20 erc20 = IERC20(tokenAddress);
-        erc20.safeTransferFrom(msg.sender, address(this), amount);
-
-        emit ERC20Deposit(depositNonce, batch.nonce);
+        if (!_isTokenMintBurn(tokenAddress)) {
+            totalBalances[tokenAddress] += amount;
+            IERC20 erc20 = IERC20(tokenAddress);
+            erc20.safeTransferFrom(msg.sender, address(this), amount);
+        } else {
+            if (!nativeTokens[tokenAddress]) {
+                require(mintBalances[tokenAddress] >= burnBalances[tokenAddress] + amount, "Not enough minted tokens");
+            }
+            burnBalances[tokenAddress] += amount;
+            IBurnableERC20 erc20 = IBurnableERC20(tokenAddress);
+            erc20.burnFrom(msg.sender, amount);
+        }
```

Comment:
This is a bit difficult to reason about. Rewriting this piece of code can help with readability. Consider this new version:

```solidity
        IERC20 erc20 = IERC20(tokenAddress);
        IBurnableERC20 burnableErc20 = IBurnableERC20(tokenAddress);

        if (_isTokenMintBurn(tokenAddress)) {
            if (!nativeTokens[tokenAddress]) {
                uint256 availableTokens = mintBalances[tokenAddress] - burnBalances[tokenAddress];
                require(availableTokens >= amount, "Not enough minted tokens");
            }
            burnBalances[tokenAddress] += amount;
            burnableErc20.burnFrom(msg.sender, amount);
        } else {
            totalBalances[tokenAddress] += amount;
            erc20.safeTransferFrom(msg.sender, address(this), amount);
        }
```

This new version:
1. Declares token interfaces at the beginning. This does not create additional operations that use gas, this type of casting is only syntactic sugar.
2. Reverses the if-else condition for better flow
3. Simplifies the nested if statement
4. Uses more descriptive variable names
5. Passes current tests
Maintains the same functionality as the original code

Reply:
done